### PR TITLE
/hotpatch chat preserves chat plugin exports

### PR DESCRIPTION
**Pull request by Morfent**
If a chat plugin exports anything besides commands, it is cached in
``CommandParser.plugins``. If the state of anything needs to be kept when
/hotpatch chat is used, the plugin can export what needs to be kept
and reassign once one of its commands is used for the first time.

### CODE REVIEW

**Comment by Slayer95** [comment]
I don't think this is a good idea... Maybe set a reserved key name so that only an specific exported property gets preserved?

File: command-parser.js [hunk 1]
```diff
@@ -44,9 +44,13 @@ if (customCommands && customCommands.commands) {
...
 
 // Install plug-in commands
-
+var plugins = exports.plugins = {};
 fs.readdirSync('./chat-plugins').forEach(function (file) {
-	if (file.substr(-3) === '.js') Object.merge(commands, require('./chat-plugins/' + file).commands);
+	if (file.substr(-3) === '.js') {
+		var plugin = require('./chat-plugins/' + file);
+		Object.merge(commands, plugin.commands);
+		if (Object.keys(plugin).length > 1) plugins[file.slice(0, -3)] = Object.reject(plugin, 'commands');
```
Comment (Slayer95) [inline command-parser.js]: What's the point of this?

**Comment by Zarel** [comment]
Shouldn't hotpatch chat hotpatch plugins too?

**Comment by Zarel** [comment]
Really, the impression I'm getting here is that the chat plugin system is a huge mess because I didn't write it. :|

File: command-parser.js (hunk 1 above)
Comment (Morfent) [inline command-parser.js]: This checks if there are other properties than ``commands`` exported by a plugin, then adds those to ``CommandParser.plugins`` under the name of the script.

**Comment by Morfent** [comment]
It does hotpatch them, but the issue is that this resets the variables that their commands use to their initial value.

What would be a better way to write the chat plugin system?

**Comment by Morfent** [comment]
This still lets hotpatch chat hotpatch the plugins while keeping any objects the plugin would need to keep, like Trivia games or Scavenger hunts. Without this, it's like hotpatching Tournaments without caching ``Tournaments.tournaments`` and setting it back afterwards, or commands if Rooms or Users had to be completely reloaded as well,

**Comment by Morfent** [comment]
It seems a bit excessive, but would splitting the plugins into two files, one for commands, and another for the heart of the plugin work? That way, command-parser.js would only have to load the command files for the plugins, while a separate module could handle loading the plugins' variables, constructors, and whatnot under separate keys in a new global, like ``ChatPlugins``. The parser could then be set up to pass each plugin's key under ``ChatPlugins`` as a parametre to the plugin's functions.

File: command-parser.js (hunk 1 above)
Comment (Morfent) [inline command-parser.js]: If someone decides to name their file just .js or give it a reserved name for some reason, it's still possible to reach the plugin's key with ``CommandParser.plugins["pluginName"]`` notation

**Comment by Morfent** [comment]
Only add plugins' ``exports.cache``, if any, to ``CommandParser.plugins`` instead of all exports besides commands.
Plugins newly added through hotpatches will no longer have their cache overwritten.

**Comment by Zarel** [comment]
>It does hotpatch them, but the issue is that hotpatching resets the plugins' variables that their commands use back to their initial value.

So, re-load them from your config file, then. Or store them in data fields attached to the room. How does tournaments do it?

**Comment by Morfent** [comment]
Tournaments exports runnning tournaments to ``Tournaments.tournaments``. /hotpatch caches this before reloading the scripts in /tournaments, which sets this back to an empty object, and reassigning its cached value so running tournaments will still continue while newly created ones use the new code.

**Comment by Morfent** [comment]
Chat plugins can now keep track of state without having to change hotpatch or how chat plugins are loaded.

**Comment by Zarel** [comment]
By "chat plugins" you just mean Trivia, don't you?

**Comment by Morfent** [comment]
No, this would also be an issue with Scavengers, Jeopardy, and Wi-Fi

**Comment by Zarel** [comment]
Yeah, but you're the only one to fix it.

**Comment by Zarel** [comment]
I mean, you only fixed Trivia.

**Comment by Morfent** [comment]
Yes, but I needed to one working for one plugin before writing one for the rest of them, over rewriting fixes of rewriting fixes for every other plugin before the fix would be up to standards.

**Pull request merged**